File: Fall 2019/CCPS 506/Assignments/Poker.ex
defmodule Poker do

    # The deal function takes the input deck,
    # deals the cards out to two hands, 
    # sorts the cards and then calls the
    # analytical functions.
    # Once a winner has been found, 
    # the format_card function is used to 
    # map the list to the correct output.   
    def deal(deck) do
        deck = Enum.take(deck, 10)

        hand1 = Stream.take_every(deck, 2) |> Stream.map(&(find_rank_and_suit(&1))) |> Enum.sort()
        hand2 = (tl deck) |> Stream.take_every(2) |> Stream.map(&(find_rank_and_suit(&1))) |> Enum.sort()

        hand1_score = check_flush(hand1) + check_straight(hand1) + check_duplicates(hand1)
        hand2_score = check_flush(hand2) + check_straight(hand2) + check_duplicates(hand2)

        winner = cond do
            hand1_score > hand2_score -> hand1
            hand1_score < hand2_score -> hand2
            hand1_score == hand2_score -> case high_card(hand1, hand2) do
                {1,0} -> hand1
                {0,1} -> hand2
                _ -> case high_suit(hand1, hand2) do
                    true -> hand1
                    false -> hand2
                end
            end
        end

        winner |> Enum.map(&(format_card(&1)))
    end

    # Function used to map card from deck to its rank and suit values
    defp find_rank_and_suit(card) do
        unless rem(card, 13) == 0 do
            {rem(card, 13), trunc(card/13)}
        else
            {13, trunc(card/13) - 1}
        end
    end


    # CHECKING FOR FLUSHNESS - Recursive
    ###################################################
    # Recursive check for flush - header w/ default args
    defp check_flush(hand, flush_suit \\ nil, is_flush \\ true)
    # Initial call to function with no suit set to check for flushness
    # Suit to check for flushness set to 2 element of first tuple in hand list
    defp check_flush(hand, flush_suit, is_flush) when is_nil(flush_suit) do
        flush_suit = elem hd(hand), 1
        check_flush(tl(hand), flush_suit, is_flush)
    end
    # If call made with is_flush boolean as false, score of 0 returned
    defp check_flush(_hand, _flush_suit, is_flush) when is_flush == false do
        0
    end
    # If the entire list has been traversed, return score of 8
    defp check_flush([], _flush_suit, _is_flush) do
        8
    end
    # Check the head of the remaining list against the
    # suit that's being tracked for flushness.
    # @is_flush will be set to false if they are unequal.
    # Call check_flush function again with tail and is_flush
    defp check_flush(hand, flush_suit, _is_flush) do
        is_flush = flush_suit == elem hd(hand), 1
        check_flush(tl(hand), flush_suit, is_flush)
    end


    # CHECKING FOR STRAIGHTNESS - Recursive
    #######################################
    # Header with default arguments
    defp check_straight(hand, acc \\ nil, is_straight \\ true)
    # Sets the accumulator when the function is first called
    defp check_straight(hand, acc, is_straight) when is_nil(acc) do
        acc = elem hd(hand), 0
        check_straight(tl(hand), acc + 1, is_straight)
    end
    # If hand is found to be not straight, return score of zero
    defp check_straight(_hand, _acc, is_straight) when is_straight == false do
        0
    end
    # Ace-low/high scenario -> Ace-high returns higher score
    defp check_straight([{13,_}], acc, _is_straight) do
        case acc do
            6 -> 6
            13 -> 7
            _ -> 0
        end
    end
    # Once the entire list has been traversed, return score
    defp check_straight([], _acc, _is_straight) do
        6
    end
    # Check if the first element in the list matches the
    # accumulator, which is being incremented by 1 with
    # each recursive call to check_straight
    defp check_straight(hand, acc, _is_straight) do
        is_straight = acc == elem hd(hand), 0
        check_straight(tl(hand), acc + 1, is_straight)
    end


    # CHECKING DUPLICATES
    ##############################
    # Pattern match possible duplicate combinations to return a score
    defp check_duplicates(hand) do
       case hand do
            # Four of a kind
            [{x,_},{x,_},{x,_},{x,_},{_,_}] -> 10 
            [{_,_},{x,_},{x,_},{x,_},{x,_}] -> 10
            # Full house
            [{x,_},{x,_},{x,_},{y,_},{y,_}] -> 9
            [{x,_},{x,_},{y,_},{y,_},{y,_}] -> 9
            # Three of a kind
            [{x,_},{x,_},{x,_},{_,_},{_,_}] -> 5
            [{_,_},{x,_},{x,_},{x,_},{_,_}] -> 5
            [{_,_},{_,_},{x,_},{x,_},{x,_}] -> 5
            # Two pairs
            [{x,_},{x,_},{y,_},{y,_},{_,_}] -> 4
            [{x,_},{x,_},{_,_},{y,_},{y,_}] -> 4
            [{_,_},{x,_},{x,_},{y,_},{y,_}] -> 4
            # One pair
            [{x,_},{x,_},{_,_},{_,_},{_,_}] -> 2
            [{_,_},{x,_},{x,_},{_,_},{_,_}] -> 2
            [{_,_},{_,_},{x,_},{x,_},{_,_}] -> 2
            [{_,_},{_,_},{_,_},{x,_},{x,_}] -> 2
            # Anything else
            [{_,_},{_,_},{_,_},{_,_},{_,_}] -> 0 
       end
    end


    # TIE BREAK WITH HIGH CARD
    # Returns 2-value tuple 
    # {1,0} - hand1 is winner
    # {0,1} - hand2 is winner
    # {0,0} - still tied, go to high_suit
    ########################################
    # Hands are ordered in ascending order, so
    # reverse to start checking in descending
    # order of rank.
    defp high_card(hand1, hand2) do
        score = {0,0}
        high_card(Enum.reverse(hand1), Enum.reverse(hand2), score)
    end
    # If all cards have been compared, return score
    defp high_card([], [], score) do
        score
    end
    # If a score has been found, return score
    defp high_card(_hand1, _hand2, score) when score != {0,0} do
        score
    end
    # If score is still tied at zero, compare ranks
    # of card tuples represented by head1/head2.
    # Set the score based on the comparison
    # and call high_card again.
    defp high_card([head1|tail1], [head2|tail2], score) when score == {0,0} do
        score = case elem(head1,0) == elem(head2,0) do
            true -> {0,0}
            false -> case elem(head1,0) > elem(head2,0) do
                true -> {1,0}
                false -> {0,1}
            end
        end
        high_card(tail1, tail2, score)
    end


    # TIE BREAK WITH HIGH SUIT
    ######################################
    # This function only runs if the 2 hands have been
    # found to be identical in rank.
    # It is only necessary to compare the suits of the 
    # highest cards of each hand. 
    # This is because two cards of identical rank MUST 
    # have different suits.
    # Returns true if hand1 has the high suit, otherwise false.
    defp high_suit(hand1, hand2) do
        elem(Enum.fetch!(hand1, 4), 1) > elem(Enum.fetch!(hand2, 4), 1)
    end


    # Function to format card tuple to string output
    defp format_card({rank, suit}) do
        suit_decoder = {"C", "D", "H", "S"}
        cond do
            rank == 13 -> Integer.to_string(1) <> elem(suit_decoder, suit)
            true -> Integer.to_string(rank + 1) <> elem(suit_decoder, suit)
        end       
    end
end

File: Fall 2019/CCPS 506/Assignments/Poker.rs
use std::collections::HashMap;

//Card and Hand structs
#[derive(Debug)]
#[derive(Eq)]
#[derive(PartialEq)]
#[derive(PartialOrd)]
#[derive(Ord)]
#[derive(Copy)]
#[derive(Clone)]
struct Card {
    rank: i32,
    suit: i32,
}

struct Hand {
    cards: [Card; 5],
    score: i32,
}

// Main function
fn main () {
}


// Functions

fn deal (deck: &[i32; 52]) -> [String; 5] {

    // Initializing Card arrays that accept the dealt cards and
    // are then used to initalize the Hand structs for analysis.
    let mut deal1: [Card; 5] = [Card {rank: -1, suit: -1,},Card {rank: -1, suit: -1,},Card {rank: -1, suit: -1,},Card {rank: -1, suit: -1,},Card {rank: -1, suit: -1,}];
    let mut deal2: [Card; 5] = [Card {rank: -1, suit: -1,},Card {rank: -1, suit: -1,},Card {rank: -1, suit: -1,},Card {rank: -1, suit: -1,},Card {rank: -1, suit: -1,}];

    // Dealing the deck, alternating by card.
    for index in 0..10 {
        if index % 2 == 0 {
            deal1[index / 2].rank = deck[index];
        } else {
            deal2[index / 2].rank = deck[index];
        }
    }    

    // Initializing Hand structs
    let mut hand1 = Hand {cards: deal1, score: 0,};
    let mut hand2 = Hand {cards: deal2, score: 0,};

    find_rank_and_suit (&mut hand1);
    find_rank_and_suit (&mut hand2);

    check_straight_and_flush (&mut hand1);
    check_straight_and_flush (&mut hand2);

    // It's impossible for a hand that is straight or flush
    // to have duplicate rank cards, so rank duplicates are
    // only checked if no score has been assigned.
    if hand1.score == 0 {       
        check_duplicates (&mut hand1);
    }
    if hand2.score == 0 {       
        check_duplicates (&mut hand2);
    }

    // After analyzing the hands
    if hand1.score == hand2.score {
        tie_break (&mut hand1, &mut hand2);
    }

    winner (&hand1, &hand2)
}

fn find_rank_and_suit (hand: &mut Hand) {

    // Processing each card of the hand
    for index in 0..5 {

        // Dividing by 13 will give a value representing the card's suit.
        // 0 = Clubs, 1 = Diamonds, 2 = Hearts, 3 = Spades
        if hand.cards[index].rank % 13 == 0 {
            hand.cards[index].suit = hand.cards[index].rank / 13 - 1; // Adjusting for ace
        } else {
            hand.cards[index].suit = hand.cards[index].rank / 13;
        }

        // The remainder will be the rank of the card.
        if hand.cards[index].rank % 13 == 0 {
            hand.cards[index].rank = 13; // Adjusting for ace
        } else {
            hand.cards[index].rank = hand.cards[index].rank % 13;
        }
    }

    // Putting the card array into sorted order based
    // on first value of each card tuple (rank value).
    hand.cards.sort();
}

fn check_straight_and_flush (hand: &mut Hand) {

    // Checking if straight and assigning a score of 
    // 6 or 7 if it is an ace-high straight.

    let mut straight_score = 6;
    // For the ace-high scenario:
    if hand.cards[0].rank == 9 && hand.cards[4].rank == 13 {
        straight_score = 7;
    }

    let mut straight_count = 4;
    // For ace-low scenario:
    if hand.cards[0].rank == 1 && hand.cards[4].rank == 13 {
        straight_count = 3;
    }

    let straight = hand.cards[0].rank + straight_count;

    if straight == hand.cards[straight_count as usize].rank {
        hand.score += straight_score;
    }

    // Checking if flush
    let flush_score = 8;
    let flush_suit = hand.cards[0].suit;
    let mut flush = true;
    let mut index = 1;

    while flush && index < 5 {
        if hand.cards[index].suit != flush_suit {
            flush = false;
        }
        index += 1;
    }

    if flush {
        hand.score += flush_score;
    }
}

fn check_duplicates (hand: &mut Hand) {

    let four_score = 10;
    let full_house_score = 9;
    let three_score = 5;
    let two_pair_score = 4;
    let pair_score = 2;

    // Use HashMap to track number of duplicate occurrences
    let mut duplicates: std::collections::HashMap<i32, i32> = HashMap::new();

    for card in &hand.cards {
        let update = duplicates.entry(card.rank).or_insert(0);
        *update += 1;
    }

    // Remove single occurrences from the map
    duplicates.retain(|_, &mut occurrences| occurrences != 1);

    // Sum the values of duplicate occurrences
    let mut duplicate_sum = 0;
    for val in duplicates.values() {
        duplicate_sum += val;
    }

    // duplicate_sum is equal to the number of cards left after discarding singles.
    // duplicates.len() gives the number of groups of duplicate cards.
    // These two values can be used to assign a score to the hand.
    hand.score += match (duplicates.len(), duplicate_sum) {
        (1, 4) => four_score,
        (2, 5) => full_house_score,
        (1, 3) => three_score,
        (2, 4) => two_pair_score,
        (1, 2) => pair_score,
        (_, _) => 0
    };
}

fn tie_break (hand1: &mut Hand, hand2: &mut Hand) {

    let tie_break_score = 1;
    let mut found = false;
    let mut end_of_index = false;
    let mut index = 4;

    // Check high card
    while !found && !end_of_index {
        let rank_difference = hand1.cards[index].rank - hand2.cards[index].rank;
        match rank_difference.signum() {
            1 => {
                hand1.score += tie_break_score;
                found = true;
            },
            -1 => {
                hand2.score += tie_break_score;
                found = true;
            },
            _ => {
                if index == 0 {
                    end_of_index = true;
                } else { 
                    index -= 1; 
                }
            },
        } 
    }

    // If no high card, reset index and check suits
    index = 4;
    while !found {
        let suit_difference = hand1.cards[index].suit - hand2.cards[index].suit;
        match suit_difference.signum() {
            1 => {
                hand1.score += tie_break_score;
                found = true;
            },
            -1 => {
                hand2.score += tie_break_score;
                found = true;
            },
            _ => {
                index -= 1;
            },
        }
    }
}

fn winner (hand1: &Hand, hand2: &Hand) -> [String; 5] {
    
    // Array to translate Card struct's suit value 
    // from integer to character representation
    let suit_decoder = ['C','D','H','S'];

    let winner = match hand1.score > hand2.score {
        true => hand1.cards,
        false => hand2.cards,
    };

    let mut output: [String; 5] = [String::new(),String::new(),String::new(),String::new(),String::new()];

    for index in 0..5 {
        if winner[index].rank == 13 {
            output[index] = format!("{}{}", 1, suit_decoder[winner[index].suit as usize]); 
        } else {
            output[index] = format!("{}{}", winner[index].rank + 1, suit_decoder[winner[index].suit as usize]);
        }
    }

    output
}